lib/srv/alpnproxy/auth/auth_proxy.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+type sitesGetter interface {
+	GetSites() ([]reversetunnel.RemoteSite, error)
+}
+
+type authGetter interface {
+	GetClusterName(opts ...services.MarshalOption) (types.ClusterName, error)
+	GetAuthServers() ([]types.Server, error)
+}
+
+// NewAuthProxyDialerService create new instance of AuthProxyDialerService.
+func NewAuthProxyDialerService(reverseTunnelServer sitesGetter, accessPoint authGetter) *AuthProxyDialerService {
+	return &AuthProxyDialerService{
+		reverseTunnelServer: reverseTunnelServer,
+		accessPoint:         accessPoint,
+	}
+}
+
+// AuthProxyDialerService allows dialing local/remote auth service base on SNI value encoded as destination auth
+// cluster name and ALPN set to teleport-auth protocol.
+type AuthProxyDialerService struct {
+	reverseTunnelServer sitesGetter
+	accessPoint         authGetter
+}
+
+func (s *AuthProxyDialerService) HandleConnection(ctx context.Context, conn net.Conn, connInfo alpnproxy.ConnectionInfo) error {
+	defer conn.Close()
+	clusterName, err := getClusterName(connInfo)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	authConn, err := s.dialAuthServer(ctx, clusterName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer authConn.Close()
+	if err := s.proxyConn(ctx, conn, authConn); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func getClusterName(info alpnproxy.ConnectionInfo) (string, error) {
+	if len(info.ALPN) == 0 {
+		return "", trace.NotFound("missing ALPN value")
+	}
+	protocol := info.ALPN[0]
+	if !strings.HasPrefix(protocol, string(alpnproxy.ProtocolAuth)) {
+		return "", trace.BadParameter("auth routing prefix not found")
+	}
+	routeToCluster := strings.TrimPrefix(protocol, string(alpnproxy.ProtocolAuth))
+	cn, err := auth.DecodeClusterName(routeToCluster)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return cn, nil
+}
+
+func (s *AuthProxyDialerService) dialAuthServer(ctx context.Context, clusterNameFromSNI string) (net.Conn, error) {
+	clusterName, err := s.accessPoint.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if clusterName.GetClusterName() == clusterNameFromSNI {
+		return s.dialLocalAuthServer(ctx)
+	}
+	if s.reverseTunnelServer != nil {
+		return s.dialRemoteAuthServer(ctx, clusterNameFromSNI)
+	}
+	return nil, trace.NotFound("auth server for %q cluster name not found", clusterNameFromSNI)
+}
+
+func (s *AuthProxyDialerService) dialLocalAuthServer(ctx context.Context) (net.Conn, error) {
+	authServers, err := s.accessPoint.GetAuthServers()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(authServers) == 0 {
+		return nil, trace.NotFound("empty auth servers list")
+	}
+	authServerIndex := rand.Intn(len(authServers))
```

Comment:
Let's leave a todo note here to better support HA in future and retry on other auth server if one fails.